table/block_based/filter_block.h

```diff
@@ -75,12 +78,28 @@ class FilterBlockBuilder {
   }
   // If filter_data is not nullptr, Finish() may transfer ownership of
   // underlying filter data to the caller,  so that it can be freed as soon as
-  // possible.
+  // possible. BlockBasedFilterBlock will ignore this parameter.
+  //
+  // If filter_data_cache_res_handle is null, Finish() may transfer
+  // ownership of cache reservation for the memory used by filter data
+  // BlockBasedFilterBlock will ignore this parameter.
+  //
+  // REQUIRED:
+  // If BlockBasedTableOptions:: reserve_table_builder_memory == true,
+  // and a block cache is availabe, that is, we charge filter construction
+  // memory, then we need the following to be true for
+  // filter_data_cache_res_handle to manage filter_data's cache reservation
+  // correctly:
+  // (filter_data && filter_data_cache_res_handle) || (!filter_data &&
+  // !filter_data_cache_res_handle)
```

Comment:
Note to @pdillinger :
We allows either `filter_data` or `filter_data_cache_res_handle` to be `nullptr` for compatibility with existing usage of `FilterBlock::Finish()`. But we have a stricter constrain on them when memory reservation is needed. 

I've been thinking of a way to write an assertion for this requirement but find it hard to access information of whether we want memory reservation or not at `FilterBlock` level.

Maybe we can do `FilterBlock::Finish(tmp, filter_data = nullptr, reserve_filter_construction_mem = false, filter_data_cache_res_handle = nullptr)` and assert the following:
`
if (reserve_filter_construction_mem) {
    assert (filter_data && filter_data_cache_res_handle) || (!filter_data && !filter_data_cache_res_handle) 
}
`

Reply:
No longer needed due to API change